src/data/multiset/basic.lean

```diff
@@ -1539,6 +1539,71 @@ theorem filter_add_not (s : multiset α) :
   filter p s + filter (λ a, ¬ p a) s = s :=
 by rw [filter_add_filter, filter_eq_self.2, filter_eq_nil.2]; simp [decidable.em]
 
+/-- Given a (decidable) property p on mutlisets and a multiset satisfying it,
+we can assume that this multiset has minimal cardinality -/
+
+lemma can_assume_min [decidable_eq α] (p : multiset α → Prop) [decidable_pred p]
+  (s₀ : multiset α) (H : p s₀) :  ∃ s ≤ s₀,  p s  ∧ (∀ a ∈ s, ¬ p (s.erase a)) :=
```

Comment:
I think this statement is not as strong as you'd like it to be.  What if `p s` is that `s.card` is odd?  Then `s₀` would satisfy the conclusion since you can't remove a single element to make the multiset smaller while having it satisfy `p`.

Here's a stronger statement that you could use to prove yours:
```lean
lemma exists_minimal_subset [decidable_eq α] (p : multiset α → Prop) [decidable_pred p]
  (s₀ : multiset α) (H : p s₀) : ∃ s ≤ s₀, p s ∧ (∀ s' ≤ s, p s' → s' = s) :=
begin
  refine multiset.strong_induction_on s₀ (λ s₀ ih H, _) H,
  by_cases h : ∃ s < s₀, p s,
  { rcases h with ⟨s₁, hs₁, ps₁⟩,
    specialize ih s₁ hs₁ ps₁,
    rcases ih with ⟨s, hsle, hs⟩,
    use ⟨s, le_of_lt (lt_of_le_of_lt hsle hs₁), hs⟩, },
  { push_neg at h,
    use [s₀, rfl.ge, H],
    intros s' hs' ps',
    cases lt_or_eq_of_le hs' with hs'' hs'',
    { exact (h _ hs'' ps').elim, },
    { exact hs'', }, },
end
```

Reply:
```lean
lemma exists_minimal_subset' [decidable_eq α] (p : multiset α → Prop) [decidable_pred p]
  (s₀ : multiset α) (H : p s₀) : ∃ s ≤ s₀, p s ∧ (∀ a ∈ s, ¬p (s.erase a)) :=
begin
  rcases exists_minimal_subset p s₀ H with ⟨s, hsle, ps, hs⟩,
  use [s, hsle, ps],
  intros a ha perase,
  have : s.erase a < s := erase_lt.mpr ha,
  rw hs (s.erase a) (erase_le a s) perase at this,
  exact lt_irrefl _ this,
end
```

Reply:
Very good point! I'll study your suggestions, but just to understand better: have you commited any change, or are you proposing that I incorporate both `exists_minimal_subset` _and_ `exists_minimal_subset'` in a new commit? Also, as far as I understand, the current version of my lemma `can_assume_min` would become obsolete, is that right?

Reply:
I haven't committed anything -- I'm just giving these to you, and you can use them however you like.

It seems to me that `exists_minimal_subset` is the more useful one, and if you need a statement like `exists_minimal_subset'`, you should instead have a lemma that if you have a minimal subset for which a predicate is true, then it's not true for any erasure:
```lean
/- not sure what to call this -/
lemma not_erase_of_minimal [decidable_eq α] (s : multiset α) (p : multiset α → Prop)
  (h : ∀ s' ≤ s, p s' → s' = s) (a : α) (ha : a ∈ s) : ¬p (s.erase a) :=
begin
  intro perase,
  have : s.erase a < s  := erase_lt.mpr ha,
  rw h _ (erase_le a s) perase at this,
  exact lt_irrefl _ this,
end
```
(Note: for `exists_minimal_subset` it turns out you can remove all the decidability constraints.)

Reply:
Ok, thanks so much. I'll work on this and turn the label back to `awaiting-review` once done and commited.

Reply:
In case you do not mind assuming `classical` facts, below is a short proof of `exists_minimal_subset`.

```
-- I could not find this in mathlib, but it seems like it could be there...
lemma nat.well_ordered_of_exists {s : set ℕ} [decidable_pred (λ w, w ∈ s)] (ex : ∃ w : ℕ, w ∈ s) :
  ∃ n ∈ s, ∀ m ∈ s, n ≤ m :=
⟨nat.find ex, nat.find_spec ex, λ m, nat.find_min' ex⟩

lemma exists_minimal_subset {p : multiset α → Prop} {s₀ : multiset α} (H : p s₀)
  [decidable_pred (λ w, w ∈ {i | ∃ u, u ≤ s₀ ∧ p u ∧ u.card = i})] :
  ∃ s ≤ s₀, p s ∧ (∀ s' ≤ s, p s' → s' = s) :=
begin
  obtain ⟨n, ⟨sm, sm0, ps, sc⟩, F⟩ := nat.well_ordered_of_exists
    (⟨_, _, le_rfl, H, rfl⟩ : ∃ w : ℕ, w ∈ { i | ∃ u, u ≤ s₀ ∧ p u ∧ u.card = i }),
  exact ⟨sm, sm0, ps, λ t tsm pt, eq_of_le_of_card_le tsm
    (by {rw sc, exact F _ ⟨t, le_trans tsm sm0, pt, rfl⟩})⟩,
end
```